test/test_main.py

```diff
@@ -0,0 +1,276 @@
+            state = self._read_and_clear_state()
+            stats['count_called'] += state.get(
+                'ConfigChangedEvent_count_called', 0)
+            stats['count_deferred'] += state.get(
+                'ConfigChangedEvent_count_deferred', 0)
+            stats['event_type_names_observed'].extend(
+                state.get('event_type_names_observed', []))
+
+        # base64 encoding is used to avoid null bytes.
+        charm_config_defer = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+            'ConfigChangedEvent_defer': True
+        }))
+
+        # First run "install" to make sure all hooks are set up.
+        update_stats(self._simulate_event('install',
+                                          charm_config_defer), stats)
+
+        self.assertEqual(stats['count_called'], 0)
+        self.assertEqual(stats['count_deferred'], 0)
+        self.assertEqual(stats['event_type_names_observed'],
+                         ['InstallEvent'])
+
+        update_stats(self._simulate_event('config-changed',
+                                          charm_config_defer), stats)
+
+        self.assertEqual(stats['count_called'], 1)
+        self.assertEqual(stats['count_deferred'], 1)
+        self.assertEqual(stats['event_type_names_observed'],
+                         ['InstallEvent', 'ConfigChangedEvent'])
+
+        charm_config_dont_defer = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+            'ConfigChangedEvent_defer': False
+        }))
+
+        # Re-emit should pick the deferred config-changed.
+        update_stats(self._simulate_event('update-status',
+                                          charm_config_dont_defer), stats)
+
+        self.assertEqual(stats['count_called'], 2)
+        self.assertEqual(stats['count_deferred'], 1)
+        self.assertEqual(stats['event_type_names_observed'],
+                         ['InstallEvent', 'ConfigChangedEvent',
+                          'ConfigChangedEvent', 'UpdateStatusEvent'])
+
+    def test_all_events_handled(self):
+        # Infer supported events from the charm class and CharmEvents.
+        charm_events = {}
+        charm_events.update({event_name: instance for event_name, instance
+                             in CharmEvents.__dict__.items()
+                             if type(instance) == Event})
+        charm_events.update({event_name: instance for event_name, instance
+                             in type(charm.Charm.on).__dict__.items()
+                             if type(instance) == Event})
+
+        logger.debug(f'Expected events {charm_events.keys()}')
+
+        charm_config = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+        }))
+
+        # First run "install" to make sure all hooks are set up.
+        self._simulate_event('install', charm_config)
+
+        # Simulate hook executions for every event.
```

Comment:
Simliarly, there's no need for this. This is testing the exact same logic every time.

Reply:
This is just to make sure that any valid event name is properly handled.

For example, we have:

- single word events: {install, start, ...};
- events with a single dash: {config-changed, upgrade-charm, leader-elected, ...};
- events with multiple dashes: {leader-settings-changed, <endpoint-name>-relation-joined, ...}

I'd skip that if we haven't had to do string manipulation ("-" -> "_" conversion) and dynamic lookup of attributes based on those strings.

All events are tested to make sure there is coverage for anything that might be added to core events in the future.

Reply:
This is not making sure that all events are tested. It's making sure that the events iterated over are tested, and you have the exact same logic both in the test and in here to iterate over the events. If this list misses half the events, nothing bad will happen. The bug goes unperceived.

Again, the test is simpler, more readable, and more realistic, if the events are hooked up as you expect real people to hook up events.

Reply:
OK, agreed.